[FIX] add prepare_experiments to fix indexing bug

mhuyck opened the pull request:
This fixes #250, resolving the bug resulting from use of model_attr='experiments__all' as
an attribute. That syntax likely never worked, but it failed silently until a recent upgrade to Haystack.

rzelayafavila commented on adage/analyze/search_indexes.py:
```diff
@@ -25,11 +25,14 @@ class SampleIndex(indexes.SearchIndex, indexes.Indexable):
         return self.get_model().objects
+
+    def prepare_experiments(self, obj):
+        return [e.accession for e in obj.experiments.all()]
```
Nice! Shall we include a unit test to make sure this method outputs what we expect it to output?

mhuyck replied:
Ohh. Yes, good idea.

- [ ] test new `prepare_experiments` method
- [ ] test to ensure future indexing problems are caught automatically

mhuyck commented:
I added a longish note to that last changeset, which I'll paste here since it's easy to miss in the GitHub UI:

> In the course of writing up a test for this, I discovered that our original implementation of the Sample index was done incorrectly. This arose from a conceptual error (on my part) in how the Haystack indexes work. The Experiment accession numbers at the heart of this fix were never getting into the "full text" index because they were not in the sample_text.txt template! The fields defined in search_indexes.py (which I thought were being searched) are only there for filtering and sorting purposes -- and we aren't currently using them at all.
> 
> This fix repairs the indexing process for the SampleIndex.experiments field in case we decide to use it. Also included is a test which causes a Haystack index to be built for Samples, which will catch an error like this earlier in the future. In addition, the full text index for Samples is now updated to include a list of corresponding Experiment accession numbers.

rzelayafavila commented on adage/analyze/templates/search/indexes/analyze/sample_text.txt:
```diff
@@ -1,4 +1,6 @@
 {{ object.name }}
 {{ object.ml_data_source }}
+Experiments: {% for e in object.experiments.all %}{{ e.accession }}{% endfor %}
+Annotations:
```
Hmmm - I'm not exactly sure how this `Experiments: ` and `Annotations: ` labels will work inside this text template, since I haven't been able to find any docs on it. Would you happen to know the location of some docs that cover this?

mhuyck replied:
I don't know of any documentation on this. The way we use these derives from the example at:

> http://django-haystack.readthedocs.io/en/master/tutorial.html#creating-searchindexes

The template comes into play in two ways:

1. the rendered template for each `Sample` is what is used to create the index with `use_template=True`, which is what we actually search via the api in `SearchResource.get_object_list` at line 69 (the part that links the template to this search is the use of the pseudo field `content`):

  ```python
  sqs = sqs.filter(content=AutoQuery(query_str)).load_all().highlight()
  ```

2. When we display the search results, the highlighted snippet that is shown is built from the text in this template.

Those two "headings" you noticed are not particularly helpful for (1), but I'm thinking they will be useful context for (2), helping the user to understand the connection between the search hit and the sample found.

I just realized that the demo site is not updated. I will do that so you can see what this looks like.

- [ ] update demo site with latest changeset

mhuyck replied:
Okay, demo site at http://greenelab.fgtech.com/ now has the latest code and has been re-indexed to show what's new.

rzelayafavila replied:
Right, ok, that makes sense @mhuyck

rzelayafavila commented on adage/analyze/tests.py:
```diff
@@ -1096,12 +1096,27 @@ class SearchIndexTestCase(ResourceTestCaseMixin, TestCase):
+        # retrieve an experiment and link some samples to test the sample index
+        exp_obj = Experiment.objects.get(pk='E-GEOD-24262')
+        for s in self.samples:
+            exp_obj.sample_set.create(**s)
```
Cool! I am not that familiar with this way of using `create` (though it is evidently using an object manager just like when we call `Sample.objects.create`). This makes it so that we don't have to pass in the desired Experiment object as an argument to link it to these new samples, huh?

mhuyck replied:
Yep! This is one of the many ways in which Django rocks.

mhuyck replied:
I probably learned about this feature through the getting started tutorial here:

> https://docs.djangoproject.com/en/1.11/intro/tutorial02/#playing-with-the-api

rzelayafavila replied:
Nice!

rzelayafavila commented on adage/analyze/tests.py:
```diff
@@ -1140,5 +1155,26 @@ def testSearchPA14(self):
+                if item['item_type'] == 'sample'    # only testing samples
+            ]),
+            frozenset([u'GSM596626 1', u'GSM596819 1'])
+        )
```
Nice! :+1: Great test!

mhuyck replied:
Thanks!

rzelayafavila commented on adage/analyze/tests.py:
```diff
@@ -1140,5 +1155,26 @@ def testSearchPA14(self):
+        accession number.
+        """
+        respEGEOD24262 = self.api_client.get(
+            self.searchURI,
```
Oh, interesting - this is definitely outside the scope of this pull request, but I forgot that we have one big search endpoint, instead of one search endpoint for each resource, or model (something like `'/api/v0/sample/search/'`). Do you think there is a better approach out of the two, or is it dependent on the project...?

mhuyck replied:
I am a fan of having one master search index that covers anything you might wish to find. If needed, filters can be applied for context-specific searches.

We don't quite do this right now because `django-genes` is a separate component, but so far don't have a use case that calls for a search that encompasses `Sample`s, `Experiment`s, *and* `Gene`s either.

rzelayafavila replied:
Ok, gotcha - yep, that makes sense with `django-genes`

mhuyck commented:
Bonus: this also fixes #233!
